fhevm/tee_comparison.go

```diff
@@ -0,0 +1,227 @@
+	// A more efficient way would be to use binary.BigEndian.UintXX().
+	// However, that would require a switch case. We prefer for now to use
+	// big.Int as a one-liner that can handle variable-length bytes.
+	//
+	// Note that we do arithmetic operations on uint64, then we convert th
+	// result back to the FheUintType.
+	l := big.NewInt(0).SetBytes(lp.Value).Uint64()
+	r := big.NewInt(0).SetBytes(rp.Value).Uint64()
+
+	result := operator(l, r)
+	resultBz, err := marshalBool(result)
+	if err != nil {
+		return nil, err
+	}
+	teePlaintext := tee.NewTeePlaintext(resultBz, 0, caller)
+
+	resultCt, err := tee.Encrypt(teePlaintext)
+	if err != nil {
+		logger.Error(op, "failed", "err", err)
+		return nil, err
+	}
+	importCiphertext(environment, &resultCt)
+
+	resultHash := resultCt.GetHash()
+	logger.Info(fmt.Sprintf("%s success", op), "lhs", lhs.hash().Hex(), "rhs", rhs.hash().Hex(), "result", resultHash.Hex())
+	return resultHash[:], nil
+}
+
+func doComparison2Operation(op string, environment EVMEnvironment, caller common.Address, input []byte, runSpan trace.Span, operator func(uint64, uint64) uint64) ([]byte, error) {
+	logger := environment.GetLogger()
+
+	lp, rp, lhs, rhs, err := extract2Operands(op, environment, input, runSpan)
+	if err != nil {
+		logger.Error(op, "failed", "err", err)
+		return nil, err
+	}
+
+	// If we are doing gas estimation, skip execution and insert a random ciphertext as a result.
+	if !environment.IsCommitting() && !environment.IsEthCall() {
+		return importRandomCiphertext(environment, lp.FheUintType), nil
+	}
+
+	// TODO ref: https://github.com/Inco-fhevm/inco-monorepo/issues/6
+	if lp.FheUintType == tfhe.FheUint128 || lp.FheUintType == tfhe.FheUint160 {
+		panic("TODO implement me")
+	}
+
+	// Using math/big here to make code more readable.
+	// A more efficient way would be to use binary.BigEndian.UintXX().
+	// However, that would require a switch case. We prefer for now to use
+	// big.Int as a one-liner that can handle variable-length bytes.
+	//
+	// Note that we do arithmetic operations on uint64, then we convert th
+	// result back to the FheUintType.
+	l := big.NewInt(0).SetBytes(lp.Value).Uint64()
+	r := big.NewInt(0).SetBytes(rp.Value).Uint64()
+
+	result := operator(l, r)
+	resultBz, err := marshalUint(result, lp.FheUintType)
+	if err != nil {
+		return nil, err
+	}
+	teePlaintext := tee.NewTeePlaintext(resultBz, lp.FheUintType, caller)
+
+	resultCt, err := tee.Encrypt(teePlaintext)
+	if err != nil {
+		logger.Error(op, "failed", "err", err)
+		return nil, err
+	}
+	importCiphertext(environment, &resultCt)
+
+	resultHash := resultCt.GetHash()
+	logger.Info(fmt.Sprintf("%s success", op), "lhs", lhs.hash().Hex(), "rhs", rhs.hash().Hex(), "result", resultHash.Hex())
+	return resultHash[:], nil
+}
+
+func doComparison3Operation(op string, environment EVMEnvironment, caller common.Address, input []byte, runSpan trace.Span, operator func(bool, uint64, uint64) uint64) ([]byte, error) {
```

Comment:
I think only cmux has 3 operands. Maybe it's not worth creating a separate function for 3 operands

Reply:
I just don't want to see the `teeCmuxRun` function getting longer...

Reply:
Why not?